### ISSUE TITLE
Suggest removing `{}` on structs with empty "bodies"

### ISSUE THREAD

**Issue by JosephTLyons** [rust-lang/rust-clippy#8591]
### What it does

Checks for struct declarations with empty bodies and suggests removing the `{}`
Checks for instantiation of structs that have empty bodies and suggests removing the `{}`

### Category

style

### Advantage

Makes the code look a little tidier.

### Drawbacks

Not sure...

### Example 1

```rust
struct Dog {}
```

Could be written as:

```rust
struct Dog;
```

### Example 2

```rust
// Assuming `struct Dog;` exists
let dog = Dog {};
```

Could be written as:

```rust
// Assuming `struct Dog;` exists
let dog = Dog;
```

**Comment by FoseFx** [comment]
Great Idea imo!

But the snippets in the second example are not equivalent as `dog` in the first snippet is a `Dog` instance, but it's an alias for the `Dog` Struct in the second snippet.

Example:
```rust

struct Strct {}
impl Strct {
    fn ok(self) {  } // requires a Strct instance
}

fn main() {
    let s = Strct {};
    s.ok();         // works

    let s = Strct;
    s.ok();         // expected value, found struct `Strct`
}

```

**Comment by Jarcho** [comment]
Just a note you should handle both `#[non_exhaustive]` structs and `#[cfg(...)]` fields. Empty tuple structs could also be given the same treatment.

**Comment by FoseFx** [comment]
I'm not sure if I understand you correctly. Do you think `#[non_exhaustive]` marked structs should not be linted? What did you have in mind regading `#[cfg(...)]`?

**Comment by Jarcho** [comment]
Ignore the part about `#[non_exhaustive]`. Just checked and it does work fully with unit structs.

For `cfg`, something like `struct Foo { #[cfg(feature = "foo")] x: u32 }` will appear as a struct with no fields to clippy unless `foo` is enabled. Dealing with this requires re-lexing the text where the `cfg` attribute may appear. It requires getting the span for all the text between the braces and then using `rustc_lexer::tokenize` on it.

**Comment by JosephTLyons** [comment]
> Great Idea imo!
> 
> But the snippets in the second example are not equivalent as `dog` in the first snippet is a `Dog` instance, but it's an alias for the `Dog` Struct in the second snippet.
> 
> Example:
> 
> ```rust
> struct Strct {}
> impl Strct {
>     fn ok(self) {  } // requires a Strct instance
> }
> 
> fn main() {
>     let s = Strct {};
>     s.ok();         // works
> 
>     let s = Strct;
>     s.ok();         // expected value, found struct `Strct`
> }
> ```

Ah ok.  I guess I should've investigated more what was happening in the second example before jumping to conclusions.  Thanks for the clarification.
